pkg/util/acrtoken/acrtoken.go

```diff
@@ -0,0 +1,113 @@
+package acrtoken
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+
+	azcontainerregistry "github.com/Azure/azure-sdk-for-go/services/containerregistry/mgmt/2019-06-01-preview/containerregistry"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/to"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/containerregistry"
+	"github.com/Azure/ARO-RP/pkg/util/compareutils"
+	"github.com/Azure/ARO-RP/pkg/util/stringutils"
+)
+
+type Manager interface {
+	Create(context.Context, *api.OpenShiftCluster) error
+	Delete(context.Context, *api.OpenShiftCluster) error
+}
+
+type manager struct {
+	env        env.Interface
+	tokens     containerregistry.TokensClient
+	registries containerregistry.RegistriesClient
+
+	scopeMap         string
+	fullRegistryName string
+}
+
+func NewManager(env env.Interface, fpAuthorizer autorest.Authorizer) Manager {
+	return &manager{
+		env: env,
+
+		tokens:     containerregistry.NewTokensClient(env.SubscriptionID(), fpAuthorizer),
+		registries: containerregistry.NewRegistriesClient(env.SubscriptionID(), fpAuthorizer),
+
+		fullRegistryName: env.ACRName() + ".azurecr.io",
+		scopeMap:         fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.ContainerRegistry/registries/%s/scopeMaps/_repositories_pull", env.SubscriptionID(), env.ACRResourceGroup(), env.ACRName()),
+	}
+}
+
+func (m *manager) tokenResourceID(name string) string {
+	return fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.ContainerRegistry/registries/%s/tokens/%s", m.env.SubscriptionID(), m.env.ACRResourceGroup(), m.env.ACRName(), name)
+}
+
+func (m *manager) generateTokenName(clusterName string) (string, error) {
+	rs, err := stringutils.RandomString("abcdefghijklmnopqrstuvwxyz", 8)
+	if err != nil {
+		return "", err
+	}
+	return clusterName[:compareutils.Min(len(clusterName), 8)] + "-" + rs, nil
+}
+
+// Create create a token on our registry
+// see https://docs.microsoft.com/en-us/azure/container-registry/container-registry-repository-scoped-permissions
+func (m *manager) Create(ctx context.Context, oc *api.OpenShiftCluster) error {
+	for _, pss := range oc.Properties.RegistryProfiles {
+		if pss.RegistryName == m.fullRegistryName {
+			// already created
+			return nil
+		}
+	}
+
+	acrTokenName, err := m.generateTokenName(oc.Name)
+	if err != nil {
+		return err
+	}
```

Comment:
this will need to be separated out so that there is a db write in between.

Reply:
I wonder whether in the end Create and Delete should be receiving an *api.RegistryProfile as an argument rather than an *api.OpenShiftCluster.